### ISSUE TITLE
feature request: support for cudnn v5 rc

### ISSUE THREAD

**Issue by fayeshine** [Theano/Theano#4342]
Anyone can help this? https://developer.nvidia.com/rdp/cudnn-download

**Comment by nouiz** [comment]
We have it here for new convolution:

https://github.com/mgermain/Theano/tree/cudnn5

It will probably be merged this week.

**Comment by sherjilozair** [comment]
@nouiz, any plans on also wrapping the the RNN implementation?

**Comment by subodhq** [comment]
From the paper the results look very impressive, can't wait for the theano integration to see it live :)

**Comment by nouiz** [comment]
Just an update, we merged that in Theano master.

We also want to wrap the RNN stuff. But we don't know when it will be done.
So if someone want to help and work on it shortly, it would be great.

**Comment by jiumem** [comment]
@nouiz I just checked the latest version of theano, seems that the RNN stuff haven't been merged yet.

**Comment by nouiz** [comment]
Just to be clear, we merged the change to support cudnn v5 convolution
change.

But we didn't started to work on wrapping the RNN stuff. But we want to do
that.

**Comment by xingdi-eric-yuan** [comment]
@nouiz any progress?

**Comment by phiber1** [comment]
Not to be a nag, but I too was wondering about the status of wrapping the native RNN support in cuDNN v5. Big performance win for theano when you do. While we're on the subject of cuDNN v5, how about a more streamlined relationship between high-level tensor types and the new tensor descriptor in cuDNN v5?

Regards,
Mark

**Comment by zpardos** [comment]
I'll second this request for cuDNN 5 rnn/lstm support. The general release is out now and by many accounts the speedup using their LSTM API is quite substantial. I'm not interested in porting over to torch at this point but am tempted just to see what the speedup is.

**Comment by Froskekongen** [comment]
Any updates on this issue would be great! And if there is a branch where this is worked on, it would be nice to get a pointer.

**Comment by carlthome** [comment]
Yes, anyone working on adding NVIDIA's proposed LSTM speedups?

**Comment by nouiz** [comment]
We don't know when someone will start to work on that. But we want to wrap it.

Do one of you want to help? This can make it appear faster.

@phiber1 what do you mean by a more streamlined version? What few people probably know is that there first "tensor" was fixed to 4d. Then when they started to make a 5d only tensor, I pushed them to stop that madness and go to nd tensor directly. This is mostly the tensor we have.

Converting in C between both descriptor is pretty fast. This isn't what is taking times.

**Comment by devrandom2** [comment]
@nouiz Do you know if anyone has started working on this yet?

**Comment by nouiz** [comment]
On the LSTM? no. The Batch normalization is close to be merged.

**Comment by carlthome** [comment]
Status on [cuDNN 5 LSTM optimizations](https://devblogs.nvidia.com/parallelforall/optimizing-recurrent-neural-networks-cudnn-5/)? I suppose it's tricky to work it into the code base for being so low-level?

- [x] OPTIMIZATION 1: COMBINING GEMM OPERATIONS
- [ ] OPTIMIZATION 2: STREAMING GEMMS
- [ ] OPTIMIZATION 3: FUSING POINT-WISE OPERATIONS
- [ ] OPTIMIZATION 4: PRE-TRANSPOSING THE WEIGHT MATRIX
- [ ] OPTIMIZATION 5: COMBINING INPUT GEMMS
and the last optimization ("Step 3: Optimization with Many Layers") of adding more layers that I don't really get because that changes the model (right?).

Anyway, I added optimization 1 to Keras (https://github.com/fchollet/keras/pull/2523) as that was straightforward. Going by Nvidia's proposed speedup table _fused point-wise operations_ should be the priority, as that's the largest performance increase. I have no clue where that should enter into Theano's code base though. @nouiz got some pointers?

**Comment by zpardos** [comment]
TF just implemented the accelerated RNN/LSTM API in cudnn v5. Not sure if return sequences are supported. 
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/cudnn_rnn

**Comment by carlthome** [comment]
https://github.com/fchollet/keras/issues/3692

**Comment by nouiz** [comment]
Just to tell that we are also finising wrapping that function in this PR:

https://github.com/Theano/Theano/pull/4915/

If all goes well, it should be merged this week.

**Comment by carlthome** [comment]
@nouiz great news!

**Comment by lamblin** [comment]
#4915 has been merged. It has been a long week.
